src/pixelator/pna/read_processing/pipeline.py

```diff
@@ -150,6 +179,63 @@ def _add_modifier(self, modifier: PairedEndModifier, stage: ModifierStage) -> No
         if stage == "post":
             self._post_modifiers.append(modifier)
 
+    def _pre_process_paired(self, reads):
+        """Pre‐process paired reads."""
+        pre_modifiers_and_steps = self._pre_modifiers + self._pre_steps
+        read1, read2 = reads
+        n_bp1 = len(read1)
+        n_bp2 = len(read2)
+        info1 = ModificationInfo(read1)
+        info2 = ModificationInfo(read2)
+
+        for step in pre_modifiers_and_steps:
+            reads = step(*reads, info1, info2)  # type: ignore
+            if reads is None:
+                break
+
+        if reads is not None:
+            read1, read2 = reads
+
+        return read1, read2, info1, info2, n_bp1, n_bp2
+
+    def _pre_process_single(self, single_read):
+        """Pre‐process single reads."""
+        pre_modifiers_and_steps = self._pre_modifiers + self._pre_steps
+        name = single_read.name
+        is_r2 = ("R2" in name) or ("/2" in name) or ("_2" in name)
+        if is_r2:
+            read1, read2 = None, single_read
+        else:
+            read1, read2 = single_read, None
+        info1 = ModificationInfo(read1) if read1 is not None else None
+        info2 = ModificationInfo(read2) if read2 is not None else None
+
+        for modifier in pre_modifiers_and_steps:
+            if isinstance(modifier, (PairedEndModifier, PairedEndStep)):
+                if isinstance(modifier, PairedEndModifierWrapper):
+                    if is_r2:
+                        out = modifier._modifier2(read2, info2)
+                        read2 = out
+                    else:
+                        out = modifier._modifier1(read1, info1)
+                        read1 = out
+                else:
+                    out = modifier(
+                        read1, read2, ModificationInfo(read1), ModificationInfo(read2)
+                    )
+                    read1, read2 = out
+            else:
+                if is_r2:
+                    out = modifier(read2, info2)  # type: ignore
+                    read2 = out
+                else:
+                    out = modifier(read1, info1)
+                    read1 = out
+            if out is None:
+                break
```

Comment:
I'm not 100% sure I follow this. Since we are in a function that should be called for single reads I would have expected it to only deal with `SingleEndModifiers`. Perhaps we could go over this together in more detail?